src/core/collection.js

```diff
@@ -464,57 +481,87 @@ export default class Collection {
     return doc[labelField.path];
   }
 
-  /**
-   * Behaves like promised-mongo's find() method except that the results are mapped to collection instances.
-   */
   find(...args) {
     const collection = this,
           db         = collection.db,
           projection = args[1],
           // extract tyranid specific options
-          tyrOpts    = extractTyranidQueryOptions(args[2]);
+          auth       = extractAuthorization(args[2]);
 
     if (projection) {
       args[1] = parseProjection(collection, projection);
     }
 
-    const cursor = db.find(...args);
-
-    /**
-     * Patch promised-mongo's connect() method to work with our async secureQuery() method.
-     * This patch is necessary because find() itself is not async.
-     */
-    cursor.connect = async function connect() {
-      if (tyrOpts.secure && !this.tyranidQueryModified) {
-
-        const securedQuery = await collection.secureQuery(
-          this.command.query,
-          'view',
-          // allow passthrough of specific user
-          // through options
-          tyrOpts.subject || tyrOpts.user || Tyr.local.user
-        );
-
-        // if the secureQuery() returns non-truthy value, that means they have no access so provide an efficient contradiction
-        this.command.query = securedQuery || { _id: null };
-        this.tyranidQueryModified = true;
+    function cursor() {
+      const cursor = db.find(...args);
+
+      hooker.hook(cursor, 'next', {
+        post(promise) {
+          console.log('next', promise);
+          const modified = promise.then(doc => doc ? new collection(doc) : null);
+          return hooker.override(modified);
+        }
+      });
+
+      hooker.hook(cursor, 'toArray', {
+        post(promise) {
+          const modified = promise.then(docs => docs.map(doc => doc ? new collection(doc) : null));
+          return hooker.override(modified);
+        }
+      });
+
+      return cursor;
+    }
+
+    if (auth) {
+      const secure = collection.secureQuery(args[0], 'view', auth);
+
+      if (secure.then) {
```

Comment:
smart, this seems like a good compromise without getting too deep in monkey patching